world/petitions/models.py

```diff
@@ -10,7 +10,87 @@
 
 from server.utils.exceptions import PayError
 from .exceptions import PetitionError
+from web.character.models import (Clue, SearchTag)
+from world.dominion.models import PlayerOrNpc
+from world.dominion.general_dominion_commands import CmdPatronage
+
+class Interest(SharedMemoryModel):
+    name = models.CharField(max_length=30)
+    wantedads = models.ManyToManyField("WantedAd", related_name="wanted_ads")
+
+
+class WantedAd(SharedMemoryModel):
+    """This is an advertisement for wanting an RP partner"""
+    MARRIAGE = 0
+    SPONSOR = 1
+    PROTEGE = 2
+    PLOT = 3
+    TIME = 4
+    MATCH_TYPES = ((MARRIAGE, "Marriage"),(SPONSOR,"Sponsor"),(PROTEGE,"Protege"),(PLOT,"Plot"),(TIME,"Time"))
+    playtime_start = models.PositiveIntegerField(default=0)
+    playtime_end = models.PositiveIntegerField(default=0)
+    prefer_weekend = models.BooleanField(default=False)
+    interests = models.ManyToManyField(Interest)
+    clues = models.ManyToManyField(Clue)
+    searchtags = models.ManyToManyField(SearchTag)
+    blurb = models.TextField()
+    marriage = models.BooleanField(default=False)
+    sponsor = models.BooleanField(default=False)
+    protege = models.BooleanField(default=False)
+    plot = models.BooleanField(default=False)
+    active = models.BooleanField(default=False)
+    owner = models.OneToOneField(PlayerOrNpc, related_name="wanted_ad")
+
+    def match(self, dompc):
```

Comment:
I'm a little concerned about the execution time of match, since this is called on every active Wanted Ad and contains nested queries with matching search tags and clues. I think it'd be better to do matches as a complex query based on everything but social rank which can't be queried, then doing that manually at the end. Maybe it wouldn't be too bad once all the Wanted Ads are in memory as a SharedMemoryModel, but I wouldn't be surprised that this could cause the game to lock up for a minute or so the first time someone did 'mm' after a reload.

Reply:
The execution time would depend mostly on how Django implements the many2many relationship between the ads and the tags/clues. If it does it in a problematic way it would be possible to just store them as a list of strings in a .db property and do a list comparison instead.

Reply:
I think the best way is to do it as a different query distinct from matches, and then append them if the ad is in that queryset. Like:

```
my_ad = self.caller.dompc.wanted_ad
ads = WantedAd.objects.filter(active=True)
if not my_ad.plot:
    plot_matches = []
else:
    plot_matches = ads.filter(Q(plot=True) & Q(clues__in=my_ad.clues.all()) | Q(searchtags__in=my_ad.searchtags.all())).distinct()
```
In `models.py` you could pass plot_matches as an arg and change the plot check to:
```
if self in plot_matches:
    matches.append(self.PLOT)
```

You could do stuff with conditional queries and annotations, but I think that's the part that matters since a nested query for each row in the table would be pretty bad news, so extracting it to only be done once would probably be all that's needed.

Reply:
That should work.